python/ray/data/_internal/logical/optimizers.py

```diff
@@ -32,20 +32,36 @@
 
 @DeveloperAPI
 def register_logical_rule(cls: Type[Rule], insert_index: Optional[int] = None):
+    if cls in _LOGICAL_RULES:
+        return
+
     if insert_index is None:
         _LOGICAL_RULES.append(cls)
     else:
         _LOGICAL_RULES.insert(insert_index, cls)
 
 
+@DeveloperAPI
+def get_logical_rules() -> List[Type[Rule]]:
+    return list(_LOGICAL_RULES)
```

Comment:
do we need the `list()` here? it's already a list right?

Reply:
It's to make a copy so that you can't edit the internal data structure.